### PR TITLE
Support common sub-expression optimization in CursorProcessorCompiler

### PR DESCRIPTION
**Pull request by frankobe**
Reuse common subexpression optimization in https://github.com/prestodb/presto/pull/14303 to avoid generating cse methods for projection and filter expressions in CursorProcessorCompiler

Initial benchmark from `CommonSubExpressionBenchmark`
```
Benchmark                                      (dictionaryBlocks)  (functionType)  (optimizeCommonSubExpression)  Mode  Cnt        Score       Error  Units
CommonSubExpressionBenchmark.ComputeRecordSet                true            json                           true  avgt   20  1254079.283 ± 14513.663  ns/op
CommonSubExpressionBenchmark.ComputeRecordSet                true            json                          false  avgt   20  1705059.614 ± 33912.064  ns/op
CommonSubExpressionBenchmark.ComputeRecordSet                true          bigint                           true  avgt   20    17707.901 ±  3226.856  ns/op
CommonSubExpressionBenchmark.ComputeRecordSet                true          bigint                          false  avgt   20    22631.380 ±  1418.222  ns/op
CommonSubExpressionBenchmark.ComputeRecordSet                true         varchar                           true  avgt   20   134511.019 ±  2928.476  ns/op
CommonSubExpressionBenchmark.ComputeRecordSet                true         varchar                          false  avgt   20   162009.251 ±  2437.937  ns/op
CommonSubExpressionBenchmark.ComputeRecordSet               false            json                           true  avgt   20  1278711.407 ± 24349.271  ns/op
CommonSubExpressionBenchmark.ComputeRecordSet               false            json                          false  avgt   20  1695407.459 ± 18522.358  ns/op
CommonSubExpressionBenchmark.ComputeRecordSet               false          bigint                           true  avgt   20    16169.200 ±   706.694  ns/op
CommonSubExpressionBenchmark.ComputeRecordSet               false          bigint                          false  avgt   20    16731.742 ±   167.647  ns/op
CommonSubExpressionBenchmark.ComputeRecordSet               false         varchar                           true  avgt   20   126666.129 ±  1589.824  ns/op
CommonSubExpressionBenchmark.ComputeRecordSet               false         varchar                          false  avgt   20   147536.758 ±  1496.214  ns/op
```

```
== RELEASE NOTES ==

General Changes
*  Add optimization for cursor projection & filter by extract and compute common subexpressions among all projections & filter first. This optimization can be turned off by session property ``optimize_common_sub_expressions``.

```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java [hunk 1]
```diff
@@ -423,7 +424,8 @@ private static MethodDefinition generateProcessMethod(
         return method;
     }
 
-    private List<MethodDefinition> generateCommonSubExpressionMethods(
+    private static List<MethodDefinition> generateCommonSubExpressionMethods(
+            Metadata metadata,
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java:428]: Is this needed?

File: presto-main/src/main/java/com/facebook/presto/sql/gen/CursorProcessorCompiler.java [hunk 2]
```diff
@@ -75,13 +94,30 @@ public void generateMethods(SqlFunctionProperties sqlFunctionProperties, ClassDe
...
         generateProcessMethod(classDefinition, projections.size());
 
-        Map<LambdaDefinitionExpression, CompiledLambda> filterCompiledLambdaMap = generateMethodsForLambda(classDefinition, callSiteBinder, cachedInstanceBinder, filter, metadata, sqlFunctionProperties, "filter");
-        generateFilterMethod(sqlFunctionProperties, classDefinition, callSiteBinder, cachedInstanceBinder, filterCompiledLambdaMap, filter);
+        ImmutableMap.Builder<LambdaDefinitionExpression, CompiledLambda> compiledLambdaMapBuilder = ImmutableMap.builder();
+        compiledLambdaMapBuilder.putAll(generateMethodsForLambda(classDefinition, callSiteBinder, cachedInstanceBinder, filter, metadata, sqlFunctionProperties, "filter"));
+
+        for (int i = 0; i < projections.size(); i++) {
+            String methodName = "project_" + i;
+            compiledLambdaMapBuilder.putAll(generateMethodsForLambda(classDefinition, callSiteBinder, cachedInstanceBinder, projections.get(i), metadata, sqlFunctionProperties, methodName));
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/CursorProcessorCompiler.java:102]: Why is this needed (handle lambda differently from before)? Ideally we'd want to reuse lambda bytecode if they are the same cross different projections. If there's any issue preventing us from doing so, we should try to fix that first.

Comment (frankobe) [reply presto-main/src/main/java/com/facebook/presto/sql/gen/CursorProcessorCompiler.java]: @rongrong This is actually the original implementation. It is changed to generate Lambda methods from projections and filter at once with `methodName=""`

Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/sql/gen/CursorProcessorCompiler.java]: This is not necessary. `LambdaBytecodeGenerator` has an API to take `List<RowExpression>`. In fact the API that takes a single `RowExpression` internally calls the other API. So we might as well change this logic to directly generate lambdas for all projections in one go.

Comment (frankobe) [reply presto-main/src/main/java/com/facebook/presto/sql/gen/CursorProcessorCompiler.java]: @rongrong Exactly, that's how it is implemented now. The comment is based on an outdated commit

File: presto-main/src/main/java/com/facebook/presto/sql/gen/CursorProcessorCompiler.java [hunk 3]
```diff
@@ -265,7 +336,79 @@ private void generateProjectMethod(
...
+            ClassDefinition classDefinition,
+            CallSiteBinder callSiteBinder,
+            CachedInstanceBinder cachedInstanceBinder,
+            Map<LambdaDefinitionExpression, CompiledLambda> compiledLambdaMap,
+            Map<Integer, Map<RowExpression, VariableReferenceExpression>> commonSubExpressionsByLevel,
+            Map<VariableReferenceExpression, CommonSubExpressionFields> commonSubExpressionFieldsMap)
+    {
+        Parameter properties = arg("properties", SqlFunctionProperties.class);
+        Parameter cursor = arg("cursor", RecordCursor.class);
+        Parameter output = arg("output", BlockBuilder.class);
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/CursorProcessorCompiler.java:351]: What is this `output`? Is this supposed to be the result of the whole expression that will be written to the output page? If that's the case, I don't think the cse functions need it.

File: presto-main/src/main/java/com/facebook/presto/sql/gen/CommonSubExpressionFields.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+import java.util.Collection;
+import java.util.Map;
+
+import static com.facebook.presto.bytecode.Access.PRIVATE;
+import static com.facebook.presto.bytecode.Access.a;
+import static com.facebook.presto.bytecode.expression.BytecodeExpressions.constantBoolean;
+import static com.facebook.presto.bytecode.expression.BytecodeExpressions.constantNull;
+
+public class CommonSubExpressionFields
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/CommonSubExpressionFields.java:33]: Maybe put this as a static class inside `CommonSubExpressionRewritter`. The class doesn't make a lot of sense standalone, especially the functions are specific to generating certain part of expression evaluation classes.

File: presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java [hunk 5]
```diff
@@ -813,23 +795,7 @@ private static void declareBlockVariables(List<RowExpression> expressions, Param
...
-        {
-            this.evaluatedField = evaluatedField;
-            this.resultField = resultField;
-            this.resultType = resultType;
-            this.methodName = methodName;
-        }
-    }
-
-    private static class FieldAndVariableReferenceCompiler
+    static class FieldAndVariableReferenceCompiler
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java:798]: This change is unnecessary right?

File: presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java (hunk 1 above)
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java:430]: I think this change can be removed, right?

File: presto-main/src/test/java/com/facebook/presto/sql/gen/TestCursorProcessorCompiler.java [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+        Map<RowExpression, VariableReferenceExpression> commonSubExpressions = commonSubExpressionsByLevel.values().stream()
+                .flatMap(m -> m.entrySet().stream())
+                .collect(toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));
+        // X+Y as CSE
+        assertEquals(1, cseFields.size());
+        VariableReferenceExpression cseVariable = cseFields.keySet().iterator().next();
+
+        RowExpression rewrittenFilter = cseCursorCompiler.rewriteRowExpressionsWithCSE(ImmutableList.of(filter), commonSubExpressions).get(0);
+
+        List<RowExpression> rewrittenProjections = cseCursorCompiler.rewriteRowExpressionsWithCSE(ImmutableList.of(ADD_X_Y_Z), commonSubExpressions);
```
Comment (rongrong) [inline presto-main/src/test/java/com/facebook/presto/sql/gen/TestCursorProcessorCompiler.java:139]: `rewriteRowExpressionsWithCSE(rowExpressions, commonSubExpressions)`? You might want to rename `rowExpressions` to `projections` to be clear.

File: presto-main/src/test/java/com/facebook/presto/sql/gen/TestCursorProcessorCompiler.java [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+                    field(0, BIGINT),
+                    field(1, BIGINT)),
+            field(2, BIGINT));
+
+    @Test
+    public void testRewriteRowExpressionWithCSE()
+    {
+        RowExpression filter = new SpecialFormExpression(AND, BIGINT, ADD_X_Y_GREATER_THAN_2);
+        ImmutableList<RowExpression> rowExpressions = ImmutableList.of(ADD_X_Y_Z, filter);
+        Map<LambdaDefinitionExpression, LambdaBytecodeGenerator.CompiledLambda> compiledLambdaMap = generateMethodsForLambda(cursorProcessorClassDefinition, callSiteBinder, cachedInstanceBinder, rowExpressions, meta, sqlFunctionProperties, "");
```
Comment (rongrong) [inline presto-main/src/test/java/com/facebook/presto/sql/gen/TestCursorProcessorCompiler.java:125]: Lambda is not needed here right?

File: presto-main/src/test/java/com/facebook/presto/sql/gen/TestCursorProcessorCompiler.java [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+import static com.facebook.presto.util.CompilerUtils.makeClassName;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static java.util.stream.Collectors.toList;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertTrue;
+
+public class TestCursorProcessorCompiler
+{
+    private static final Metadata meta = createTestMetadataManager();
```
Comment (rongrong) [inline presto-main/src/test/java/com/facebook/presto/sql/gen/TestCursorProcessorCompiler.java:72]: Let's call the variable `METADATA` to be consistent with the codebase.

File: presto-main/src/test/java/com/facebook/presto/sql/gen/TestCursorProcessorCompiler.java [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+    private static final FunctionManager FUNCTION_MANAGER = meta.getFunctionManager();
+    private static final CursorProcessorCompiler cseCursorCompiler = new CursorProcessorCompiler(meta, true);
+
+    private static final ClassDefinition cursorProcessorClassDefinition = new ClassDefinition(
```
Comment (rongrong) [inline presto-main/src/test/java/com/facebook/presto/sql/gen/TestCursorProcessorCompiler.java:76]: All static final variables should be UPPER_CASE.

Comment (rongrong) [reply presto-main/src/test/java/com/facebook/presto/sql/gen/TestCursorProcessorCompiler.java:76]: I think these variables can be defined within `testRewriteRowExpressionWithCSE` though.

File: presto-main/src/test/java/com/facebook/presto/sql/gen/TestCursorProcessorCompiler.java [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+    }
+
+    @Test
+    public void testCompilerWithCSE()
+    {
+        PageFunctionCompiler functionCompiler = new PageFunctionCompiler(meta, 0);
+        ExpressionCompiler expressionCompiler = new ExpressionCompiler(meta, functionCompiler);
+
+        RowExpression filter = new SpecialFormExpression(AND, BIGINT, ADD_X_Y_GREATER_THAN_2, ADD_X_Y_LESS_THAN_10);
+        List<? extends RowExpression> projections = createIfProjectionList(5);
```
Comment (rongrong) [inline presto-main/src/test/java/com/facebook/presto/sql/gen/TestCursorProcessorCompiler.java:155]: Would be interesting to make the filter and projection share some cse.

File: presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java (hunk 1 above)
Comment (frankobe) [reply presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java:430]: @rongrong it is removed already

File: presto-main/src/test/java/com/facebook/presto/sql/gen/TestCursorProcessorCompiler.java (hunk 10 above)
Comment (frankobe) [reply presto-main/src/test/java/com/facebook/presto/sql/gen/TestCursorProcessorCompiler.java:150]: @rongrong mk sense. add `ADD_X_Y` to the projections so "X+Y" is shared between projections & filter
